zcash_client_backend/src/fees/common.rs

```diff
@@ -11,16 +14,131 @@ use zcash_primitives::{
 };
 use zcash_protocol::ShieldedProtocol;
 
-use crate::data_api::WalletMeta;
+use crate::data_api::{InputSource, WalletMeta};
 
 use super::{
-    sapling as sapling_fees, ChangeError, ChangeValue, DustAction, DustOutputPolicy,
-    EphemeralBalance, SplitPolicy, TransactionBalance,
+    sapling as sapling_fees, ChangeError, ChangeStrategy, ChangeValue, DustAction,
+    DustOutputPolicy, EphemeralBalance, SplitPolicy, TransactionBalance,
 };
 
 #[cfg(feature = "orchard")]
 use super::orchard as orchard_fees;
 
+/// A change strategy that attempts to split the change value into some number of equal-sized notes
+/// as dictated by the included [`SplitPolicy`] value.
+pub struct CommonChangeStrategy<I, FR> {
+    fee_rule: FR,
+    change_memo: Option<MemoBytes>,
+    fallback_change_pool: ShieldedProtocol,
+    dust_output_policy: DustOutputPolicy,
+    split_policy: SplitPolicy,
+    _meta_source: PhantomData<I>,
+}
+
+impl<I: InputSource, FR: FeeRule + Clone> CommonChangeStrategy<I, FR> {
+    /// Constructs a new [`CommonChangeStrategy`] with the specified fee rule, change memo,
+    /// dust output policy, and change splitting policy.
+    ///
+    /// This change strategy will fall back to creating a single change output if insufficient
+    /// change value is available to create notes with at least the minimum value dictated by the
+    /// split policy.
+    ///
+    /// * `fallback_change_pool`: the pool to which change will be sent if when more than one
+    ///   shielded pool is enabled via feature flags, and the transaction has no shielded inputs.
+    /// * `split_policy`: A policy value describing how the change value should be returned as
+    ///   multiple notes.
+    pub fn new(
+        fee_rule: FR,
+        change_memo: Option<MemoBytes>,
+        fallback_change_pool: ShieldedProtocol,
+        dust_output_policy: DustOutputPolicy,
+        split_policy: SplitPolicy,
+    ) -> Self {
+        Self {
+            fee_rule,
+            change_memo,
+            fallback_change_pool,
+            dust_output_policy,
+            split_policy,
+            _meta_source: PhantomData,
+        }
+    }
+
+    /// Constructs a new [`CommonChangeStrategy`] with the specified fee rule, change memo,
+    /// and a single change output. This is equivalent to [`CommonChangeStrategy::new`] with
+    /// `dust_output_policy == DustOutputPolicy::default()` and
+    /// `split_policy == SplitPolicy::single_output()`.
+    pub fn simple(
```

Comment:
```suggestion
    pub fn single_output(
```
and affected call sites